sharedmodels/sharedmodels/tests/unit/rdf/test_resource_serialisation.py

```diff
@@ -0,0 +1,268 @@
+                PropertyStatus.mandatory,
+                map_str_to_en_literal,
+            ),
+        ]
+
+    b = B("http://some-b-uri")
+    b.p = "Hello, World"
+
+    graph = b.to_graph(Graph())
+
+    assert 1 == len(graph)
+    assert (
+        URIRef("http://some-b-uri"),
+        URIRef("http://replacement-uri"),
+        Literal("Hello, World", "en"),
+    ) in graph
+    assert (
+        URIRef("http://some-b-uri"),
+        URIRef("http://original-uri"),
+        Literal("Hello, World", "en"),
+    ) not in graph
+
+
+def test_resource_serialised_for_new_resource():
+    """Ensure that we can use an `NewResource` when providing a `Resource` for serialisation."""
+
+    class A(NewResource):
+        p_a: Annotated[
+            str,
+            Triple(
+                URIRef("http://some-predicate-a"),
+                PropertyStatus.optional,
+                map_str_to_en_literal,
+            ),
+        ]
+
+    class B(NewResource):
+        p_b: Annotated[
+            Resource[A],
+            Triple(
+                URIRef("http://some-predicate-b"),
+                PropertyStatus.mandatory,
+                map_resource_to_uri,
+            ),
+        ]
+
+    a = A("http://some-new-resource-a")
+    a.p_a = "Hello, World"
+
+    b = B("http://some-new-resource-b")
+    b.p_b = a
+
+    graph = b.to_graph(Graph())
+
+    assert (
+        URIRef("http://some-new-resource-b"),
+        URIRef("http://some-predicate-b"),
+        URIRef("http://some-new-resource-a"),
+    ) in graph
+
+    # Ensure that we recurviely serialise new resources and their triples.
+    assert (
+        URIRef("http://some-new-resource-a"),
+        URIRef("http://some-predicate-a"),
+        Literal("Hello, World", "en"),
+    ) in graph
+
+
+def test_resource_serialised_for_existing_resource():
+    """Ensure that we can use an `ExistingResource` inplace of a `NewResource` when providing a `Resource` for serialisation."""
+
+    class A(NewResource):
+        p_a: Annotated[
+            str,
+            Triple(
+                URIRef("http://some-predicate-a"),
+                PropertyStatus.optional,
+                map_str_to_en_literal,
+            ),
+        ]
+
+    class B(NewResource):
+        p_b: Annotated[
+            Resource[A],
+            Triple(
+                URIRef("http://some-predicate-b"),
+                PropertyStatus.mandatory,
+                map_resource_to_uri,
+            ),
+        ]
+
+    b = B("http://some-new-resource-b")
+    b.p_b = ExistingResource("http://some-existing-resource-a")
+
+    graph = b.to_graph(Graph())
+
+    assert (
+        URIRef("http://some-new-resource-b"),
+        URIRef("http://some-predicate-b"),
+        URIRef("http://some-existing-resource-a"),
+    ) in graph
+
+
+def test_resource_serialised_for_circular_reference():
+    """
+    Ensure that we can correctly serialise graph structures which contain circular references.
```

Comment:
Take two objects `A` and `B`

- if `A` has a property which references `B` then that is okay
- if `B` *also* has a property which references `A` then that causes a circular reference.

It means that when we serialise the class to RDF we could end up in an infinite loop. This test ensures that we don't do that.

We could end up in an infinite loop since when we serialise `A` we look at all of its properties and define the relationship to `B`, but we realise that `B` is also a `NewResource` which should be serialised, so we output the triple with its relationship to `A`, but then we realise that `A` is a `NewResource` and should be serialised, and then we have an infinite loop... (unless we do something to stop that happening).